### DISCUSSION TITLE
Uncaught TypeError: jsPsych.data.get is not a function

### DISCUSSION THREAD

**Discussion by ThibaultC123** [jspsych/jsPsych#1896] [Q&A]
Hello,

I have to modify an existing jsPsych experiment to summarize reaction times over conditions.

Wherever I insert any of these lines (close to that in the manual):

`var myData = jsPsych.data.get().filter({trial_type: 'poldrack-single-stim'}).filterCustom(function(x){return x.trial_num > 0});
`

`var myData = jsPsych.data.get()`

I get the following message again and again in my console: "Uncaught TypeError: jsPsych.data.get is not a function"

Any help would be greatly appreciated. Thanks in advance.

**Comment by becky-gilbert** [comment]
Hi @ThibaultC123, that's odd... `jsPsych.data.get()` should definitely be a function, at least in newer jsPsych releases (after 6.0). Do you know what version of the jspsych.js file you're using? If you're not sure, one thing you could try is searching for `module.get` in your jspsych.js file - with jsPsych v6.0 and higher, you should find this:

```js
 module.get = function() {
    return allData;
  };
```

Also, do you see any errors in the console that occur _before_ the "Uncaught TypeError: jsPsych.data.get is not a function" error?

**Reply by ThibaultC123** [reply]
Hi @becky-gilbert ,

Thank you so much for your quick answer which is part of the solution.

I don't know which but this was an old version of jspsych that I updated. Now in the new jsPsych.js file I get the message from line 144 : "The display_element specified in jsPsych.init() does not exist in the DOM." and from line 158 : "Uncaught TypeError: Cannot read property 'indexOf' of undefined".

Hope you can help me further.

**Reply by becky-gilbert** [reply]
Can you share your `jsPsych.init()` code? Or maybe share a link to the experiment, if it's hosted somewhere? 

It sounds like the [`display_element` parameter](https://www.jspsych.org/core_library/jspsych-core/#parameters_9) is specified but that element isn't included on the HTML page. If that's the case then I would try removing the `display_element` parameter from `jsPsych.init`, which means that jsPsych will default to using the \<body> element to display the experiment.

You said you're modifying an existing experiment - do you know what jsPsych version it was initially written with? In general I think you might keep running into errors like this if the jsPsych files (jspsych.js and plugins) aren't the versions that were used when the experiment was written, because even different minor jsPsych releases can cause compatibility issues. But if you're not sure we can still try to help you try to figure out the errors as they come up 😃

**Reply by ThibaultC123** [reply]
Hello @becky-gilbert,

I run the project locally so it is not hosted anywhere.

You will find below the `jsPsych.init()` code (in the index.html):

```
<script>
    $(document).ready(function () {

        jsPsych.init({
            timeline: attention_network_task_experiment,
            display_element: "getDisplayElement",
            fullscreen: true,
            on_trial_finish: function (data) {
                addID('attention-network-task')
            },

            on_finish: function (data) {

                // Serialize the data
                var promise = new Promise(function (resolve, reject) {
                    var data = jsPsych.data.dataAsJSON();
                    resolve(data);
                })

                promise.then(function (data) {


                    $.ajax({
                        type: "POST",
                        url: '/save',
                        data: { "data": data },
                        success: function () { document.location = "/next" },
                        dataType: "application/json",

                        // Endpoint not running, local save
                        error: function (err) {

                            if (err.status == 200) {
                                document.location = "/next";
                            } else {
                                // If error, assue local save
                                jsPsych.data.localSave('attention-network-task_results.csv', 'csv');
                            }
                        }
                    });
                })
            }
        });
    });
</script>
```

I also attach the link to the github repository of the project that I have been inspired by, and that I have modified very little for the moment: https://github.com/expfactory-experiments/attention-network-task

I searched all the files, but unfortunately I could not find the version of jsPsych used in this project.

Thanks in advance.

**Reply by ThibaultC123** [reply]
Still trying to understand, bear with me. 

A simple way is to state `display_element: undefined` in jspsych.init(). 

Doing so, I now get this error: "Uncaught TypeError: Cannot read property 'parameters' of undefined" in jspsych.js at line 1082 (evaluateFunctionParameters).

**Reply by ThibaultC123** [reply]
In the initial version, there is ```module.evaluateFunctionParameters = function(trial, protect) {}``` and in version 6.3.0 it was changed to ```function evaluateFunctionParameters(trial){}```. 

Does it mean that the code of the experiment itself has to be updated too ?

**Reply by becky-gilbert** [reply]
The experiment factory task you linked to was last updated 4 years ago, which means that it was probably created with a jsPsych release from around 2016-2017. There have been some pretty big changes to jsPsych since then, so you will probably keep running into compatibility problems if you try to use this code with a more recent jsPsych release. So if you don't want to make this code compatible with recent jsPsych versions, then another option is to try using your code with a jsPsych release from around that time, e.g. v5.0.3.

> A simple way is to state display_element: undefined in jspsych.init().

If you set the `display_element` parameter to `undefined`, then you'll actually override the default display element (the page's body) with `undefined`, which will cause errors. So instead, I suggest you try removing the `display_element` parameter entirely from `jsPsych.init`. Also, the `fullscreen` option has been removed from `jsPsych.init` because it's been replaced with the [fullscreen plugin](https://www.jspsych.org/plugins/jspsych-fullscreen/) (see the [docs for jsPsych.init parameters](https://www.jspsych.org/core_library/jspsych-core/#parameters_9)). So you can remove the `fullscreen: true` line from `jsPsych.init` (and if you want the experiment to be in fullscreen, then add a fullscreen trial to the main timeline). Removing those two parameters will mean that your `jsPsych.init` code looks like this:
```js
jsPsych.init({
  timeline: attention_network_task_experiment,
  on_trial_finish: function (data) {
  // etc. ...
```
I also noticed that, in your `on_finish` function, the jsPsych data functions you have are out of date (see the [jsPsych data docs](https://www.jspsych.org/core_library/jspsych-data/)). In your code, the jsPsych data is converted to JSON format with this line: `var data = jsPsych.data.dataAsJSON();`, but now this should be written as:
```js
var data = jsPsych.data.get().json();
```
And to save the data locally, instead of `jsPsych.data.localSave('attention-network-task_results.csv', 'csv');`, you should use this:
```js
jsPsych.data.get().localSave('csv','attention-network-task_results.csv');
```

**Reply by ThibaultC123** [reply]
Thank you so much for your very helpful answers. I applied all your advices. I would like to keep with version 6.3.0 although I understand that going back to 5.0.3 would be easier.

I am running into compatibility issues. I rewrote some of the plugins using the up to date syntax. 

Now the next error in jspsych-poldrack-text.js is : "jsPsych.pluginAPI.evaluateFunctionParameters is not a function". However, this function is correctly defined.

```
plugin.trial = function(display_element, trial) {
    // if any trial variables are functions
    // this evaluates the function and replaces
    // it with the output of the function
    trial = jsPsych.pluginAPI.evaluateFunctionParameters(trial);
... 
}
```

**Reply by becky-gilbert** [reply]
You can remove that line from the plugin, because more recent versions of jsPsych automatically evaluate the trial's function parameters in the jspsych.js trial [here](https://github.com/jspsych/jsPsych/blob/ae332b33477d686388bafbd03b0c8e1c9945999c/jspsych.js#L979), before the plugin runs the trial.

**Reply by ThibaultC123** [reply]
Sorry to bother. Still progressing, but got stuck with "Uncaught TypeError: v.toLowerCase is not a function" in jspsych.js (version 6.3.0)... How can it be?

```
var valid_response = false;
      if (typeof parameters.valid_responses === 'undefined'){
        valid_response = true;
      }
      else if(parameters.valid_responses == jsPsych.ALL_KEYS) {
        valid_response = true;
      } 
      else if(parameters.valid_responses != jsPsych.NO_KEYS){
        if(parameters.valid_responses.includes(e.key)){
          valid_response = true;
        }
        if(!case_sensitive) {
          var valid_lower = parameters.valid_responses.map(function(v) { return v.toLowerCase(); });
          var key_lower = e.key.toLowerCase();
          if (valid_lower.includes(key_lower)) {
            valid_response = true;
          }
        }
      }
```

**Reply by becky-gilbert** [reply]
This might be due to the switch from using key codes to key characters (i.e. the keyboard event's `key` property - explained in the [v6.3.0 release notes](https://github.com/jspsych/jsPsych/releases/tag/v6.3.0)). So if any of the key choice parameters in your code use numbers, then you'll want to change these to characters. E.g., `choices: [70,74]` should now be `choices: ['f','j']`, and `choices: ['space']` / `choices: [32]` should now be `choices: [' ']`. This is true for all key choice parameters, including e.g. `key_forward` and `key_backward` in the instructions plugin.

**Reply by ThibaultC123** [reply]
You were right once more!

This is the last problem I bumped in after clearing a couple of others...
```
var post_trial_gap = function() {
	var curr_trial = jsPsych.progress().current_trial_global
       return 3500 - jsPsych.data.getData()[curr_trial - 1].rt - jsPsych.data.getData()[curr_trial - 4].block_duration
}
````

The aim is to compute the post_trial_gap as a function of the random duration of a fixation (i.e. trial-4) and the reaction time (i.e. trial-1). Since jsPsych.data.getData() doesn't exist any more in v6.3.0, Do you have any idea ?

**Reply by ThibaultC123** [reply]
I got it! Here is the updated code
```
var post_trial_gap = function() {	// Modified by Thibault for JsPsych 6.3.0
	var rt = jsPsych.data.getLastTrialData().select('rt').values[0]
	var D1 = jsPsych.data.get().last(4).select('block_duration').values[0]
	return 3500 - D1 - rt
}
```

**Reply by becky-gilbert** [reply]
Awesome, glad you figured it out!

**Reply by ThibaultC123** [reply]
Thankl you. Still have some pending issues with Timeout when presenting the stimulus... Also have to work on the output of results. Will let you know soon if I'm stuck. :)
